pkg/sdkserver/sdkserver.go

```diff
@@ -824,3 +830,19 @@ func (s *SDKServer) updateConnectedPlayers(ctx context.Context) error {
 	s.recorder.Event(gs, corev1.EventTypeNormal, "PlayerCount", fmt.Sprintf("Set to %d", gs.Status.Players.Count))
 	return nil
 }
+
+// NewSDKServerContext returns a Context that cancels when SIGTERM or os.Interrupt
+// is received and the GameServer's Status is shutdown
+func (s *SDKServer) NewSDKServerContext(ctx context.Context) context.Context {
+	sdkCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-ctx.Done()
+		for {
+			gsState := <-s.gsStateChannel
+			if gsState == agonesv1.GameServerStateShutdown {
+				cancel()
+			}
+		}
+	}()
+	return sdkCtx
+}
```

Comment:
🤔 I feel like this would work, and be a bit more concise - if the parent context cancels, it should flow down into the child.

```suggestion
// NewSDKServerContext returns a Context that cancels when SIGTERM or os.Interrupt
// is received and the GameServer's Status is shutdown
func (s *SDKServer) NewSDKServerContext(ctx context.Context) context.Context {
	ctx, cancel := context.WithCancel(ctx)
	go func() {
		for {
			gsState := <-s.gsStateChannel
			if gsState == agonesv1.GameServerStateShutdown {
				cancel()
			}
		}
	}()
	return ctx
}
```

Reply:
i tried this but in the event that the parent context is cancelled(main signal handler), the child will be cancelled as well(and not wait for shutdown state)🤔

Reply:
aaah yes, of course. Good point. In which case, let's get this merged 👍🏻